backends/gstreamer/src/player.rs

```diff
@@ -488,29 +495,29 @@ impl GStreamerPlayer {
                 }
 
                 let sender_clone = sender.clone();
-
-                let need_data_id = Arc::new(Mutex::new(None));
-                let need_data_id_clone = need_data_id.clone();
-                *need_data_id.lock().unwrap() = Some(
-                    appsrc
-                        .connect("need-data", false, move |args| {
-                            let _ = sender_clone.lock().unwrap().send(Ok(()));
-                            if let Some(id) = need_data_id_clone.lock().unwrap().take() {
-                                glib::signal::signal_handler_disconnect(
-                                    &args[0].get::<gst::Element>().unwrap(),
-                                    id,
-                                );
-                            }
-                            None
-                        })
-                        .unwrap(),
-                );
-
-                let inner_clone = inner_clone.clone();
+                let is_ready_ = is_ready_clone.clone();
+                let inner_ = inner_clone.clone();
+                let inner__ = inner_clone.clone();
+                let inner___ = inner_clone.clone();
                 appsrc.set_callbacks(
                     AppSrcCallbacks::new()
+                        .need_data(move |_, _| {
+                            // We block the caller of the setup method until we get
+                            // the first need-data signal, so we ensure that we
+                            // don't miss any data between the moment the client
+                            // calls setup and the player is actually ready to
+                            // get any data.
+                            if !is_ready_.load(Ordering::Relaxed) {
```

Comment:
use a Once for this

https://doc.rust-lang.org/nightly/std/sync/struct.Once.html

Reply:
(this doesn't use Relaxed, though -- any reason why Relaxed is okay in this code? You may get double initialization)